File: minecraftmod/fabric-1.20.1/src/main/clojure/cn/li/fabric1201/client/render/block_entity_renderer_impl.clj
(ns cn.li.fabric1201.client.render.block-entity-renderer-impl
  "Fabric 1.20.1 Universal Block Entity Renderer

  Platform-specific BlockEntityRenderer implementation that dispatches to registered renderers.
  This class knows nothing about specific blocks - it's a pure dispatcher."
  (:require [cn.li.mcmod.client.render.tesr-api :as tesr-api]
            [cn.li.mcmod.util.log :as log]))

(gen-class
  :name cn.li.fabric1201.client.render.BlockEntityRendererImpl
  :implements [net.minecraft.client.renderer.blockentity.BlockEntityRenderer]
  :prefix "renderer-"
  :constructors {[] []})

(defn renderer-render
  "Main render method - called by Minecraft every frame"
  [_ block-entity partial-ticks pose-stack buffer-source packed-light packed-overlay]
  (try
    (tesr-api/render-tile-entity block-entity partial-ticks pose-stack buffer-source packed-light packed-overlay)
    (catch Exception e
      (log/error "Error rendering BlockEntity in Fabric:" (.getMessage e))
      (.printStackTrace e))))

File: minecraftmod/fabric-1.20.1/src/main/java/cn/li/fabric1201/client/render/BlockEntityRendererImpl.java
package cn.li.fabric1201.client.render;

import clojure.java.api.Clojure;
import clojure.lang.IFn;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.world.level.block.entity.BlockEntity;

/**
 * Explicit Java BER dispatcher for Fabric scripted block entities.
 */
public final class BlockEntityRendererImpl implements BlockEntityRenderer<BlockEntity> {

    private static volatile IFn renderTileEntityFn;

    private static IFn renderFn() {
        IFn fn = renderTileEntityFn;
        if (fn == null) {
            synchronized (BlockEntityRendererImpl.class) {
                fn = renderTileEntityFn;
                if (fn == null) {
                    IFn require = Clojure.var("clojure.core", "require");
                    require.invoke(Clojure.read("cn.li.mcmod.client.render.tesr-api"));
                    fn = Clojure.var("cn.li.mcmod.client.render.tesr-api", "render-tile-entity");
                    renderTileEntityFn = fn;
                }
            }
        }
        return fn;
    }

    @Override
    public void render(BlockEntity blockEntity,
                       float partialTick,
                       PoseStack poseStack,
                       MultiBufferSource bufferSource,
                       int packedLight,
                       int packedOverlay) {
        try {
            renderFn().invoke(blockEntity, partialTick, poseStack, bufferSource, packedLight, packedOverlay);
        } catch (Throwable t) {
            System.err.println("[my_mod] Fabric BlockEntityRendererImpl.render failed:");
            t.printStackTrace();
        }
    }
}
